### PR TITLE
Warn if underscore character is used in hostname

### PR DESCRIPTION
**Pull request by davet2001**
# What does this implement/fix? 
Warns about underscores in hostnames to prevent DHCP/DNS problems.

 - Underscores in device names can cause strange DNS problems that are hard for users to trace, [example1](https://github.com/esphome/issues/issues/1123), [example2](https://community.home-assistant.io/t/nodes-created-from-esphome-not-showing-a-entities-in-home-assist/228807).

 - According to [RFC1912](https://datatracker.ietf.org/doc/html/rfc1912), underscores are not allowed in hostnames.

This PR causes the compile to raise an warning when the user tries to use an underscore in a hostname.

It is not a breaking change because it is just a warning. If the user choose to rename, they will have to remove & re-add the device in HA, the warning links to instructions.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Other

**Related issue or feature (if applicable):** 
Partial fix for https://github.com/esphome/issues/issues/1123

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** 
https://github.com/esphome/esphome-docs/pull/1366

## Test Environment
- [ ] ESP32
- [ ] ESP8266

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).
(The [documentation](https://esphome.io/components/esphome.html#configuration-variables) is already correct as it does not state that underscores are allowed)

### CODE REVIEW

**Review by jesserockz — APPROVED** [review]
This has been discussed a few times in the past and we removed the ability for users to set an underscore via the wizard.

File: esphome/config_validation.py [hunk 1]
```diff
@@ -876,8 +876,8 @@ def hostname(value):
     if len(value) > 63:
         raise Invalid("Hostnames can only be 63 characters long")
     for c in value:
-        if not (c.isalnum() or c in "_-"):
-            raise Invalid("Hostname can only have alphanumeric characters and _ or -")
+        if not (c.isalnum() or c in "-"):
+            raise Invalid("Hostname can only have alphanumeric characters and -")
```
Comment (agners) [inline esphome/config_validation.py:881]: For completeness, can you also check the start/end restriction, e.g.
```suggestion
            raise Invalid("Hostname can only have alphanumeric characters and -")
    if not value[0].isalnum() or not value[-1].isalnum():
        raise Invalid("Hostname must start and end with a alphanumeric character")
```

**Review by agners — APPROVED** [review]
We use the host name in DNS systems so it make sense following the DNS restrictions on hostnames :+1:

**Comment by OttoWinter** [comment]
Previous discussion here: https://github.com/esphome/esphome/pull/1632

TL;DR - I don't think we should make this breaking change.

Main things:
* I believe there are a lot of users with underscores in their node names, because some time ago that was the _only_ delimiter allowed (dash was not allowed).
* Upgrading to the new style is a big PITA and multi-step process (not just remove+readd to HA; but for OTA you also need to change `use_address` to point to the old name, then upload; then change `use_address` back, then upload again; and I'm sure there are lots of edge cases). Unless we make that easier I'm firmly against making it a hard requirement
* New users are already being guided to use dashes, and existing users presumably are not experiencing the mentioned DHCP/DNS problem. For them this would only be a PITA with no added value.

I'm now more ok with having it be a warning, but if we do so we need to add a FAQ entry with 1. why it's recommended and 2. detailing the name change process

**Comment by agners** [comment]
>     * Upgrading to the new style is a big PITA and multi-step process (not just remove+readd to HA; but for OTA you also need to change `use_address` to point to the old name, then upload; then change `use_address` back, then upload again; and I'm sure there are lots of edge cases). Unless we make that easier I'm firmly against making it a hard requirement

Renaming on ESPHome side was quite easy by changing yaml then use `esphome upload --device oldname.tld config/mycfg.yaml`. HA side was a bit more annoying. I wish we could change hostname for a active ESPHome integration.

Even tough a particular router might work with host names with underlines, there might be other gremlins waiting when using underlines. I think at least a warning would be good.

**Comment by OttoWinter** [comment]
> Renaming on ESPHome side was quite easy by changing yaml then use esphome upload --device oldname.tld config/mycfg.yaml.

Yeah from CLI it's easier - unfortunately I think most users use the dashboard, where we can't easily do that :shrug:

**Comment by davet2001** [comment]
A warning would be perfect in my opinion.  For context my scenario was:

1. Have an existing esphome device that works
2. Create a new esphome device by copying and pasting the yaml
3. Choose a new device name without realising the dangers of '_'
4. Sensor entities mysteriously all unavailable after HA restarts (for several weeks).
5. Find [someone else reporting esphome DHCP issues](https://github.com/esphome/issues/issues/1123) by accident.
6. Tail the `named` log to catch the problem.
7. Realise the dangers of '_' and fix it.

I think 5 and 6 are extremely unlikely for the average user to do.  At that point it just seemed really unreliable and I was guessing why with wide scope (including HA, wifi signal strength, sensor reliability, etc).

I don't think blocking via the wizard is enough.  Once you have a working device, the fastest way to set up a new one (wifi credentials etc) is to copy it.

How do we make the compile generate a warning?  Maybe the warning could have a link for how to rename it.  Even better, could it be underlined as a warning in the editor?

**Comment by glmnet** [comment]
I moved all my devices from `_` to `-` mostly because of debugging an issue in HA where ESPHome api services didn't work when the node name contained a `-`. Anyway didn't find it funny to do the migration. A warning seems more appropriate.

Off:
A better warning / info would be nice, also if they could be disabled (the warnings). Been thinking this for a while

**Comment by OttoWinter** [comment]
> How do we make the compile generate a warning? 

The only way right now is to print a warning at validation time [like this](https://github.com/esphome/esphome/blob/2419bc367820b985d153d7ffdc304949cc242adb/esphome/components/rtttl/__init__.py#L58-L58)

> Maybe the warning could have a link for how to rename it. 

Yes, we should do that; On the docs FAQ page I think we could add an entry (or maybe if you find a better page to put it on there)

> Even better, could it be underlined as a warning in the editor?

Unfortunately, no. There's no system right now to propagate tagged data

**Comment by davet2001** [comment]
All feedback incorporated now.  Just generates a warning, no error.
The 'breaking change' can be removed.
Note: It looked like the existing hostname validator wasn't actually being used at all; I fixed that in config.py but someone should check that part in particular!

File: esphome/config_validation.py [hunk 2]
```diff
@@ -873,11 +873,20 @@ def validate_bytes(value):
...
     for c in value:
-        if not (c.isalnum() or c in "_-"):
-            raise Invalid("Hostname can only have alphanumeric characters and _ or -")
+        if not (c.isalnum() or c in "-_"):
+            raise Invalid("Hostname can only have alphanumeric characters and -")
+        if c in "_" and not warned_underscore:
+            _LOGGER.warning(
+                "'%s': Using the '_' (underscore) character in the hostname is discouraged "
+                "as it can cause problems with some DHCP and local name services. "
+                "To rename an existing device see https://esphome.io/components/esphome.html#changing-esphome-node-name",
```
Comment (OttoWinter) [inline esphome/config_validation.py]:
```suggestion
                "For more information, see https://esphome.io/guides/faq.html#why-shouldn-t-i-use-underscores-in-my-device-name",
```

**Comment by OttoWinter** [comment]
Thanks a lot @davet2001 !
